### ISSUE TITLE
contiguous_container constructors differ signficiantly from `std::vector`

### ISSUE THREAD

**Issue by gnzlbg** [everard/contiguous-container#10]
Among the example use-cases of `contiguous_container` are the implementation of an `inline_vector`, `constexpr_vector`, and the possibility of implementing `small_vector` is also suggested in the paper.

However, `contiguous_container` inherits all its constructors from `Storage`, but there are no requirements on which constructors `Storage` needs to offer. 

So currently, depending on how `Storage` is specified, `contiguous_container` _might_ support:
- aggregate initialization (or not), 
- direct initialization from an initializer_list, e.g., `constexpr_vector<5> a({1, 2, 3});` static_assert(a.size() == 3);` "could" work, depending on how clever the `constexpr` implementation of `Storage` is,
- `some_vector a(size);` could have default or value initialization semantics (or not be available at all), ...

Having to add all/most vector constructors to `Storage` amounts to more boiler-plate than everything else combined (since some constructors like `vector(first, last)` amount to e.g. reimplementing `insert`).

I cannot really give you any constructive feedback about how to solve this issue. My constraints are: 

- I don't want to have to reimplement all `std::vector` constructors for every `Storage` type,
- I still want "hard" things like `constexpr_vector<5> a({1, 2, 3});` to work out-of-the-box.

The only way in which I could make something like this work would be to optionally allow `Storage` to define a static function for initialization from an initializer list, and to detect that using traits or inside `contiguous_container` like this:

```c++
template <typename Storage>
struct contiguous_container : Storage {

constexpr
contiguous_container(std::initializer_list<value_type> il) 
  : Storage(std::move(il)) 
  requires(has_initializer_list_constructor_v<Storage>) 
{}

constexpr
contiguous_container(std::initializer_list<value_type> il) 
  requires(not has_initializer_list_constructor_v<Storage>) 
{
  // fallback:
  assign(std::move(il)); // or similar
}

//...
};
```

but doing this for most `std::vector` constructors (or all constructors that make sense) would probably significantly complicate the implementation and specification of `contiguous_container` and the `storage_traits` =/ I wish somebody would find a nice way to solve this.

**Comment by everard** [comment]
Well, constructors are actually more simple to implement, since you know your container is empty when constructor starts working, so there is no need to reimplement insert. You can use for_each with construct for ForwardIterator case (initializer_list case is included), and emplace_back in a loop for InputIterator case. So no, you definitely don't need to reimplement all vector functionality for constructors, they are in fact trivially implementable.

And yes, I couldn't find an easy way to solve this, so I made them inherited.

**Comment by gnzlbg** [comment]
> So no, you definitely don't need to reimplement all vector functionality for constructors, they are in fact trivially implementable.

Even when they are trivially implementable, implementing these adds some boilerplate:

- `Storage(size_type N)`
- `Storage(size_type N, const T& v)`
- `Storage(InputIt first, InputIt last)`
- `Storage(std::initializer_list<T> il)`

so it would be nice if we could figure out a way to achieve that (I think that is a goal worth pursuing, I'll try to think about it). 

There are some `contiguous_container`s for which this gets a bit hard, for example if the `value_type` is `const` (e.g. `const int`, or even worse, `const MoveOnly`), and it can get a bit messier if one wants to support `constexpr_vector<const MoveOnly, 10>` for example. 

> You can use for_each with construct for ForwardIterator case (initializer_list case is included), and emplace_back in a loop for InputIterator case.

Are you talking about offering these inside `contiguous_container` or in `Storage`? `Storage` doesn't implement `emplace_back` and implementing what you suggest here kind of amounts to implementing insert for the case in which the container is empty, which is indeed as you say easier, but can still get a bit messy. 

I wonder if maybe lifting `emplace_back` from `contiguous_container` to `Storage` would allow implementing these constructors in `contiguous_container` alone.

**Comment by everard** [comment]
I understand what you are saying, but I think the issue is deeper here.
Basically, contiguous_container doesn't know what non-static data members Storage class might have, what constructors Storage might have. So contiguous_container doesn't know how to properly initialize Storage. So I don't see how to define constructors inside contiguous_container without imposing additional limitations to Storage.

Considering emplace_back, no, I used wrong words here. What I was trying to say is that emplace_back is really trivial, so you just unwrap it inside Storage. Or, actually, its implementation can be moved to storage_traits.

**Comment by gnzlbg** [comment]
> So I don't see how to define constructors inside contiguous_container without imposing additional limitations to Storage.

In my re-implementation I had to add a `Storage(std::initializer_list<T>)` constructor to support `constexpr_vector<...> a = { ... }` but I was able to move all other constructors to `contiguous_container`. 

> So `contiguous_container` doesn't know how to properly initialize Storage

Most std library types model `SemiRegular`, but currently `contiguous_container` is not even guaranteed to model `DefaultConstructible` (it only models it if `Storage` model it). 

I think it would be worth it to try to make `contiguous_storage` always model `SemiRegular`. This would impose "additional limitations" to `Storage`: it would need to be `DefaultConstructible`, `CopyAssignable`, and `CopyConstructible`. 

I think requiring `Storage` to model `DefaultConstructible` alone would be enough to solve most of the constructor issues, since that would allow `contiguous_container` constructors to first default construct the storage, and then use the `Insertable` requirement to do the rest. 

Do you know any good example of possible `Storage` types that cannot model `SemiRegular` or `DefaultConstructible` ?

**Comment by everard** [comment]
I don't know what SemiRegular is.
But maybe some small program for embedded device wants to have a vector, which is only one of a kind in program, access to which is implemented as Scott Meyer's singleton and at construction it takes address of storage (which might change depending on some factors). And its constructor is private, so nobody else constructs it. Also it has deleted copy/move constructors and assignment operators.

**Comment by gnzlbg** [comment]
cppreference has a definition of `SemiRegular` in [this page](http://en.cppreference.com/w/cpp/language/constraints) although one needs to search for it, it is a concept of the Ranges TS. It basically means "this type behaves similar to built-in types like `int`" which allows the user to transfer their intuition to your type.  

>But maybe some small program for embedded device wants to have a vector, which is only one of a kind in program, access to which is implemented as Scott Meyer's singleton and at construction it takes address of storage (which might change depending on some factors). And its constructor is private, so nobody else constructs it. 

I see, but if you cannot add a static member function to `contiguous_storage` that returns a `contiguous_storage&` (as opposed to `Storage&`), preventing others from creating multiple objects of that type becomes quite hard, and impossible (AFAIK) without writing a wrapper. A user would be probably better off by just writing a small wrapper like this:

```c++
template <typename T>
struct unique {
  private:
    T value_;
    unique() {}
    // deleted copy/move construction/assignment
  public:
    static T& get() {
      static unique<T> v();
      return v;
    } 
};

// And be done with it:
bool e = unique<contiguous_container<Storage>>::get().empty();
```

**Comment by everard** [comment]
I actually agree with you, and I understand your concern, but I still think contiguous_container is not the right place to define constructors. The purpose of this facility is to make it easier to define other vector-like classes without imposing unnecessary limitations. It is up to users of this facility to decide how to use it.

Also a set of constructors might be defined in another derived-from-Storage class. We can impose aforementioned limitations on that class and pass it to the contiguous_container, which will then simply inherit constructors.

So basically `my_storage -> storage_wrapper<my_storage> -> contiguous_container<storage_wrapper<my_storage>>`

and this storage_wrapper is something like
```
template<typename S>
struct storage_wrapper : protected S {
  // all the needed constructors are defined here
};
```

How about this?

**Comment by gnzlbg** [comment]
> How about this?

That might be a nice solution.

**Comment by everard** [comment]
I'm closing this issue. Rationale is the following: some container classes need to be able to inherit _from_ contiguous_container. One such example: in order for [fixed_vect](https://github.com/KevinDHall/Embedded-Containers) by Kevin Hall to be implementable in terms of contiguous_container, contiguous_container must be able to have no constructors. In which case actual container class (ssf_vect and rsf_vect in this case) inherits from contiguous_container and provides its own constructors. Nonetheless, something like storage_wrapper must be defined, there is no doubt about it.

**Comment by gnzlbg** [comment]
> Rationale is the following: some container classes need to be able to inherit from contiguous_container. One such example: [...] fixed_vect 

Why can't fixed_vect be implemented as `fixed_vect = contiguous_container<SomeStorageType>;`? 

This is actually interesting, since doing so is kind of the point of `contiguous_container`.

> One such example: in order for fixed_vect by Kevin Hall to be implementable in terms of contiguous_container, contiguous_container must be able to have no constructors. 

I have an implementation of a constexpr fixed vector using something like contiguous container, and the only constructor I cannot implement in contiguous container is the initializer list one (which I actually implement in contiguous container as well, but need to forward it to the one in the storage implementation.

I would like to know which exact constructors cannot be implemented by contiguous container for fixed vector and why. If it is only the initializer list constructor, it might still be worth it to provide the other three in `contiguous_container`. While I kind of like `storage_wrapper` solution better, I again wonder if it is really necessary. I'll try to play a bit more with all it and see if I can come with good and bad examples of it (I haven't implemented `storage_wrapper` yet).

**Comment by everard** [comment]
> I would like to know which exact constructors cannot be implemented by contiguous container for fixed vector and why

None of them, because Kevin Hall's fixed_vect is a base class for ssf_vect and rsf_vect, which contain non-static data members. And fixed_vect itself contains pointer, which is set to allocated inside ssf_vect or rsf_vect memory on construction of said classes.

So basically:
```
template <typename T>
struct fixed_vect_storage
{
        using value_type = T;

        constexpr T* begin() noexcept
        {
                return storage_;
        }

        constexpr const T* begin() const noexcept
        {
                return storage_;
        }

        constexpr auto& size() noexcept
        {
                return size_;
        }

        constexpr auto& size() const noexcept
        {
                return size_;
        }

        constexpr auto capacity() const noexcept
        {
                return capacity_;
        }

protected:
        T* storage_{};
        std::size_t capacity_{}, size_{};
};
template <typename T>
using fixed_vect = ecs::contiguous_container<fixed_vect_storage<T>>;

template <typename T, std::size_t N>
struct ssf_vect : fixed_vect<T>
{
        ssf_vect()
        {
                this->storage_ = reinterpret_cast<T*>(data_);
                this->capacity_ = N;
        }

       // other constructors here

private:
        alignas(T) unsigned char data_[N * sizeof(T)]{};
};
```

**Comment by everard** [comment]
As a side note: storage_wrapper must detect if Storage is allocator-aware (by checking for existence of allocator_type), and if it does, then it should provide constructors, which take additional allocator parameter, just like std::vector does.

**Comment by gnzlbg** [comment]
Why is `alignas(T) unsigned char data_[N * sizeof(T)]{};` in `ssf_vect` instead of in `fixed_vect_storage` ? That kind of makes `ssf_vect` the storage... it is weird.

**Comment by gnzlbg** [comment]
> As a side note: storage_wrapper must detect if Storage is allocator-aware (by checking for existence of allocator_type), and if it does, then it should provide constructors, which take additional allocator parameter, just like std::vector does.

This can be done in a `contiguous_container` constructor as well:

```c++
// only exists if `Storage` is `AllocatorAware`:
contiguous_container(std::size_t n, allocator_t<Storage> alloc) 
  requires(AllocatorAware<Storage>) { ... }
```

**Comment by gnzlbg** [comment]
I would implement `ssf_vect` as follows:

```c++
template <typename T, std::size_t N>
struct fixed_vect_storage
{
        using value_type = T;

        constexpr T* begin() noexcept
        {
                return data_;
        }

        constexpr const T* begin() const noexcept
        {
                return data_;
        }

        constexpr auto& size() noexcept
        {
                return size_;
        }

        constexpr auto& size() const noexcept
        {
                return size_;
        }

        constexpr auto capacity() const noexcept
        {
                return N;
        }

        // realloc handles size + data

protected:
        alignas(T) unsigned char data_[N * sizeof(T)]{};
        std::size_t size_{};
};
template <typename T, std::size_t N>
using ssf_vect = ecs::contiguous_container<fixed_vect_storage<T, N>>;
```

**Comment by gnzlbg** [comment]
It seems that he wants to use fixed vector in interfaces without using the `std::size_t N` parameter on the type. 

I wonder why he doesn't just use a `fixed_vector_view<T>` class for that instead. Using inheritance like that seems to unnecessarily complicate the design, since his fixed_vector type is both a vector and a view of the vector. 

The proper way to design this, is to make them different types, with different tasks. Doing it this way you can solve the problem in two different ways, which are both better than his:

- using virtual dispatch (like he does), but more flexible, so that beyond `fixed_vector<T, N>` for different `N`, the class also works with `std::vector`, `boost::vector`, `string`, ... all can be supported. 
- without virtual dispatch (faster than his way, allows inlining), but restricted to knowing the vector types you want to support in advance (e.g. a closed set of values of N, and a closed set of `std::vector`, `boost::vector, ...). 

You can do both by reusing your `contiguous_container`class, if you define the `Storage` type as follows:

```c++
template <typename T>
struct vector_view_storage {

template <typename Vector>
vector_view_storage(Vector&& v) 
  // since this is a view, `Vector` cannot be a temporary
  requires(not RValueReference<Vector>)
  : data_([&v]() { return v.data(); ) 
  : size_([&v]() { return v.size(); ) 
  : capacity_([&v]() { return v.capacity(); ) 
{}

  // Note: std::function is used for exposition only to
  // explain "the essence" of the trick.
  // A zero-cost implementation of this (without std::function) is available in
  // Sean Parent's Concept-Based run-time Polymorphism talks.
  function<T*()> data_;
  function<std::size_t()> size_;
  function<std::size_t()> capacity_;

  // implement data/begin/end/realloc in term of data_, size_, capacity_.
};

template <typename T>
struct any_vector_view = contiguous_container<vector_view_storage<T>>;
```

That has "view" / "reference" semantics, but it is trivial to make a `vector_view_storage` type that has value semantics (it owns the vector), and it is also trivial to remove virtual dispatch (which his implementation has) by just using a `variant<fixed_vector<T, 1>, fixed_vector<T,2>, ..., std::vector<T, polymorphic_allocator>, boost::vector<...>, ...>` for the storage, if you know how many vector types you have to support in advance (which makes it _faster_ than his type).

**Comment by everard** [comment]
This `variant<fixed_vector<T, 1>, fixed_vector<T,2>, ...` is not acceptable, because we don't know the capacity, we need to write variant for vectors with all possible capacities, then.

The key issue here is binary bloat. Will any_vector_view reduce it? Or will it make it worse?

**Comment by gnzlbg** [comment]
> Will any_vector_view reduce it? Or will it make it worse?

An `any_vector_view` that uses virtual functions internally has the same overhead code-wise than a base class that has virtual functions internally. The only difference is that it works with anything that looks like a vector, and not just a concrete `fixed_vector` type. 

> This variant<fixed_vector<T, 1>, fixed_vector<T,2>, ... is not acceptable, because we don't know the capacity, we need to write variant for vectors with all possible capacities, then. 

Nothing is free. You can choose between infinite inflexibility and virtual dispatch (which prevents inlining in 99% of the situations), or finite flexibility and inlining. 

While that variant can be generated automatically for a maximum `N` (e.g. by using `std::make_integer_sequence`) , `N` would still need to be fixed at compile time. 

Whether one of the versions (finite+inlining vs infinite+virtual) is better than the other depends on the situation. If you hardcode one as a base class, you get the worst of all worlds. You loose inlining, and you loose the possibility of passing other vector types in your interfaces.

By having them as different types you can choose: is this function in the hot path and you can live with a maximum `N` of 250 or 500 or whatever? Use `variant_vector_view<T, 512>` to get the variant version that works with all fixed_vectors from N = 1 to N = 512 (that's a lot of vectors). You don't care and want infinite flexibility? Use `any_vector_view<T>`, which even works on `vartiant_vector_view<T, N>` types..

**Comment by gnzlbg** [comment]
I think that the problem he wants to solve is worth solving. The fact that fixed vector types with different capacities are incompatible is a source of pain, in particular on APIs. But there are better ways of solving this problem, and these ways don't have the interactions with contiguous_container than his solution does.

**Comment by everard** [comment]
virtual dispatch is a no-go. Also generating variant for all possible Ns is a bad idea, I think, because it will have to instantiate N templates, hence bloat. Memory is essential for small devices, so if all that template magic bloats binary, then it will not be used.

**Comment by gnzlbg** [comment]
I just read Kevin Hall's implementation and he doesn't use virtual dispatch (I just assumed he did). 

So what about adding a `fixed_vector_view<T>` like this:

```c++
template <typename T>
struct fixed_vector_view_storage {

template <typename T, std::size_t N>
explicit fixed_vector_view_storage(ssf_vector<T, N>& v) :
  data_(v.data()),
  size_(v.size()),
  capacity_(v.capacity()) {}

  T* data_;
  size_t size_;
  size_t capacity_;

  // begin/end/data/reallocate
};

template <typename T>
using fixed_vector_view = contiguous_container<fixed_vector_view_storage<T>>;
```

That doesn't require virtual functions, nor more template instantiations than his solution, and doesn't add extraneous constructor requirements.

EDIT: but this would require public inheritance of the `Storage` type to support construction from a different type.

**Comment by everard** [comment]
Viable, but now you have to pass additional parameter to all constructors, because fixed_vector_view_storage takes another storage as a parameter. And how do you define contiguous_container's constructors?

**Comment by gnzlbg** [comment]
One way would be to do it like this:

```c++
template <typename Storage>
struct contiguous_container : Storage {

contiguous_container(std::size_t n, const value_type& v) 
  requires(DefaultConstructible<Storage>)
  : Storage() {
    traits<Storage>::reallocate(this, n);
    auto ptr = data();
    while(n > 0) {
      traits<Storage>::construct(this, pos, v);
      --n;
      ++ptr;
  }
}
};
```

since that would only enable the constructor if `Storage` is `DefaultConstructible`, which in this case it isn't. This could be an example of a situation for which it really makes sense to not require `Storage` to be `DefaultConstructible`. 

But I don't know if this is the _best_ way to do it, it is only one way.

In particular, I don't think we can call `contiguous_container<fixed_vector_view_storage<T>>` a container, since it doesn't own its elements (as opposed to other containers). Note that on destruction of the "view", the elements of the container are not destroyed, because the "view" (like `fixed_vector<T>&/const&` in Kevin Hall's implementation) does not own its elements.

Maybe it would be better to call it `vector_facade` (like the `iterator_facade` and `view_facade`/`range_facade` proposed in the Ranges TS) because what it does is make something look like a vector. The main difference, however, is the inversion of inheritance in these facade classes. They work like this: 

```c++
template <typename T, std::size_t N>
struct my_vector_type_storage 
  : vector_facade<my_vector_type_storage<T, N>> {
    // stuff goes here
}
```

so basically it is the _user_ of `vector_facade` (i.e. the derived type) who decides whether it wants to inherit the methods of `vector_facade` publicly, protected, or privately, and whether it wants to wrap them or not. This design, which is a bit different, might turn to be more flexible and solve all our problems, so I guess it could be worth exploring.

**Comment by everard** [comment]
vector_facade seems like a nice name.
But I can't understand your example:
```
template <typename T, std::size_t N>
struct my_vector_type_storage 
  : vector_facade<my_vector_type_storage<T, N>> {
    // stuff goes here
}
```
Why base class list contains mention of my_vector_type_storage? Is it similar to curiously recurring template? Also what stuff goes there? Constructors? And what do we do next with this my_vector_type_storage? And how do we end up with ssf_vect from this point?
